src/common/url.cpp

```diff
@@ -4,19 +4,33 @@
 
 #include <common/url.h>
 
-#include <event2/http.h>
-
-#include <cstdlib>
 #include <string>
+#include <stdexcept>
+#include <charconv>
 
-std::string urlDecode(const std::string &urlEncoded) {
+std::string UrlDecode(const std::string& url_encoded) {
     std::string res;
-    if (!urlEncoded.empty()) {
-        char *decoded = evhttp_uridecode(urlEncoded.c_str(), false, nullptr);
-        if (decoded) {
-            res = std::string(decoded);
-            free(decoded);
+    res.reserve(url_encoded.size());
+
+    for (size_t i = 0; i < url_encoded.size(); ++i) {
+        char c = url_encoded[i];
+        // Special handling for percent which should be followed by two hex digits
+        // representing an octet values, see RFC 3986, Section 2.1 Percent-Encoding
+        if (c == '%' && i + 2 < url_encoded.size()) {
+            int decoded_value = 0;
+            auto [ptr, ec] = std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16);
+            if (ec == std::errc()) {
+                res += static_cast<char>(decoded_value);
+                // Next two characters are part of the percent encoding
+                i += 2;
+            } else {
+                // In case of invalid percent encoding, add the '%' and continue
+                res += c;
```

Comment:
we could unify these two `res += c` cases by merging the two conditions:
```C++
std::string UrlDecode(const std::string& url_encoded) {
    std::string res;
    res.reserve(url_encoded.size());

    for (size_t i = 0; i < url_encoded.size(); ++i) {
        int decoded_value = 0;
        if (url_encoded[i] == '%' &&
            i + 2 < url_encoded.size() &&
            std::from_chars(&url_encoded[i + 1], &url_encoded[i + 3], decoded_value, 16).ec == std::errc{}) {
            res += static_cast<char>(decoded_value);
            i += 2;
        } else {
            res += url_encoded[i];
        }
    }

    return res;
}
```

Reply:
I think this is less readable so I will keep it as is but I did take @Sjors 's suggestion to change it to change `ptr` to `_`.

Reply:
Like above, I think this is a bit less readable so I will keep it as is.